### PR TITLE
GH-79, add rpm virtual repository

### LINKED ISSUE #79: Virtual repository key pair assignement does not work
**Describe the bug**
Changing/assigning `key_pair` on a virtual repository has no effect. Apply never converges.
The issue has been fixed here: https://github.com/atlassian/go-artifactory/pull/31
As this repository has not much activity for a while. Can jfrog take ownership or fork that repository so that we can use it in this repository ?

terraform 0.15
artifactory 7.17.4

```
resource "artifactory_virtual_repository" "repo" {
  key          = "repo"
  package_type = "rpm"
  key_pair     = "key1"
}
```

### PR DESCRIPTION
**Pull request by danielmkn**
Add new resource `artifactory_virtual_rpm_repository` with support of `primary_keypair_ref` and `secondary_keypair_ref`

Fixes #79

### CODE REVIEW

File: pkg/artifactory/repositories.go [hunk 1]
```diff
@@ -597,24 +601,28 @@ var baseVirtualRepoSchema = map[string]*schema.Schema{
...
+		Optional:    true,
+		Default:     false,
+		Description: "Whether the virtual repository should search through remote repositories when trying to resolve an artifact requested by another Artifactory instance.",
 	},
 	"default_deployment_repo": {
-		Type:     schema.TypeString,
-		Optional: true,
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Default repository to deploy artefacts.",
```
Comment (alexhung) [inline pkg/artifactory/repositories.go:625]: Typo with `artefacts`

File: pkg/artifactory/resource_artifactory_virtual_rpm_repository.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		Type:             schema.TypeString,
+		Optional:         true,
+		ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotEmpty),
+		Description:      "Secondary keypair used to sign artifacts.",
+	},
+})
+
+type CommonRpmDebianVirtualRepositoryParams struct {
+	PrimaryKeyPairRef   string `hcl:"primary_keypair_ref" json:"primaryKeyPairRef,omitempty"`
+	SecondaryKeyPairRef string `hcl:"primary_keypair_ref" json:"secondaryKeyPairRef,omitempty"`
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_virtual_rpm_repository.go:26]: Same question (always), do we need `omitempty`? I doubt we need it so let's always start without.

File: pkg/artifactory/resource_artifactory_virtual_rpm_repository.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+func resourceArtifactoryRpmVirtualRepository() *schema.Resource {
+	return mkResourceSchema(rpmVirtualSchema, defaultPacker, unpackRpmVirtualRepository, func() interface{} {
+		return &RpmVirtualRepositoryParams{
+			VirtualRepositoryBaseParams: VirtualRepositoryBaseParams{
+				Rclass:      "virtual",
+				PackageType: "rpm",
+			},
+		}
+	})
+
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_virtual_rpm_repository.go:43]: Let's tidy up all these extra empty lines.

**Review by alexhung — CHANGES_REQUESTED** [review]
Apart from the very minor typo and empty lines, LGTM.

**Review by alexhung — CHANGES_REQUESTED** [review]
Apart from the very minor typo and empty lines, LGTM.

File: pkg/artifactory/resource_artifactory_virtual_repository_test.go [hunk 4]
```diff
@@ -250,6 +250,138 @@ func TestAccVirtualMavenRepository_basic(t *testing.T) {
 	})
 }
 
+func TestAccVirtualRpmRepository(t *testing.T) {
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_virtual_repository_test.go:253]: There is a ticket about support for GPG keys somewhere - and I thought I hadn't implemented them. Maybe I did. Go find that ticket and see if you can close it. It's pretty recent

File: pkg/artifactory/resource_artifactory_virtual_rpm_repository.go (hunk 2 above)
Comment (chb0github) [reply pkg/artifactory/resource_artifactory_virtual_rpm_repository.go:26]: Please assume you don't need it. Also, is there a reason no one is using the built in functionality to general the hcl and then using a tag? Usage of `hcl:` was only needed before I implemented auto-detect. If you add it, it's an override.

Comment (chb0github) [inline pkg/artifactory/resource_artifactory_virtual_rpm_repository.go:26]: You have `hcl:primary_keypair_ref` twice. Another reason to just remove it

File: pkg/artifactory/resource_artifactory_virtual_rpm_repository.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	repo := RpmVirtualRepositoryParams{
+		VirtualRepositoryBaseParams: unpackBaseVirtRepo(s, "rpm"),
+		CommonRpmDebianVirtualRepositoryParams: CommonRpmDebianVirtualRepositoryParams{
+			SecondaryKeyPairRef: d.getString("secondary_keypair_ref", true),
+			PrimaryKeyPairRef:   d.getString("primary_keypair_ref", true),
+		},
+	}
+	repo.PackageType = "rpm"
+
+	return &repo, repo.Key, nil
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_virtual_rpm_repository.go:58]: This could be simplified to just return an `Identifiable` - but, meh...

**Review by chb0github — CHANGES_REQUESTED** [review]
You have an actual bug

File: pkg/artifactory/resource_artifactory_virtual_repository_test.go (hunk 4 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_virtual_repository_test.go:253]: This one? https://github.com/jfrog/terraform-provider-artifactory/issues/285

Comment (chb0github) [reply pkg/artifactory/resource_artifactory_virtual_repository_test.go:253]: So, GPG is supported? I remember there is no way to clientside validate it There needs to be a test just for GPG support for #285 - if memory serves correct, 2 reasons I couldn't support it were: 
1. no client side validation and
2. Couldn't get RT to accept all the test keys I tried.

**Review by chb0github — APPROVED** [review]
It's approved with the caveat that GPG is an actual supported key type

File: pkg/artifactory/resource_artifactory_virtual_repository_test.go (hunk 4 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_virtual_repository_test.go:253]: The tests are using GPG type and they pass, so RT is accepting them. I can add GPG tests to `artifactory_keypair`

**Review by chb0github — APPROVED** [review]

**Review by alexhung — APPROVED** [review]

**Pull request merged**